ML project/Customer_churn_prediction.py

```diff
@@ -0,0 +1,512 @@
+lr.fit(X_train,y_train)
+preds_lr = lr.predict(X_test)
+preds_prob_lr = lr.predict_proba(X_test)[:,1]
+fpr_lr, tpr_lr, tresholds_lr = roc_curve(y_test,preds_prob_lr)
+prec_lr, recall_lr, tresh_lr = precision_recall_curve(y_test,preds_prob_lr)
+auc_score_lr = roc_auc_score(y_test,preds_prob_lr)
+
+
+# ## RandomForest
+
+# In[37]:
+
+
+from sklearn.ensemble import RandomForestClassifier
+rf = RandomForestClassifier()
+
+
+# In[38]:
+
+
+params = {'n_estimators' : [300,400,500],
+          'max_depth': [15,20,30],
+          'min_samples_leaf':[2,3]}
+
+for i in ['f1','balanced_accuracy','accuracy','roc_auc']:
+    rcv = RandomizedSearchCV(rf, param_distributions = params, cv=3,scoring = i,random_state=123)
+    rcv.fit(X_train,y_train)
+    print('Metrics: {}\nBest params: {}\nBest score: {}'.format(i,rcv.best_params_,rcv.best_score_))
+
+
+# In[39]:
+
+
+f1_rf = []
+b_acc_rf = []
+acc_rf = []
+auc_rf = []
+
+for i in ['f1','balanced_accuracy','accuracy','roc_auc']:
+    rf = RandomForestClassifier(n_estimators = 400, min_samples_leaf = 2, max_depth = 20)
+    metrics = cross_val_score(rf, X_train, y_train, cv = 10, scoring=i)
+    if i == 'f1':
+        f1_rf.append(metrics)
+    elif i == 'balanced_accuracy':
+        b_acc_rf.append(metrics)
+    elif i == 'roc_auc':
+        auc_rf.append(metrics)
+    else:
+        acc_rf.append(metrics)
+
+
+# In[40]:
+
+
+plt.bar(['F1','Balanced accuracy','AUC','Accuracy'],[np.mean(f1_rf), np.mean(b_acc_rf), np.mean(auc_rf),np.mean(acc_rf)])
+plt.yticks([0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1])
+plt.title('RF Metric score (average)');
+
+
+# In[41]:
+
+
+rf = RandomForestClassifier(n_estimators = 400, min_samples_leaf = 2, max_depth = 20)
+rf.fit(X_train,y_train)
+preds_rf = rf.predict(X_test)
+preds_prob_rf = rf.predict_proba(X_test)[:,1]
+fpr_rf, tpr_rf, tresholds_rf = roc_curve(y_test,preds_prob_rf)
+prec_rf, recall_rf, tresh_rf = precision_recall_curve(y_test,preds_prob_rf)
+auc_score_rf = roc_auc_score(y_test,preds_prob_rf)
+
+
+# In[42]:
+
+
+fi_df = pd.DataFrame(rf.feature_importances_, index = df_scaled.drop('churn',axis=1).columns, columns = ['feature_importance'])
+fi_df[fi_df.feature_importance>0.03].sort_values('feature_importance').plot(kind='barh');
+
+
+# In[43]:
+
+
+fig, [ax1,ax2] = plt.subplots(1,2)
+fig.set_size_inches(10,5)
+
+ax1.plot([0,1],'k--')
+ax1.plot(fpr_k_8,tpr_k_8,'c',label = 'KNC')
+ax1.plot(fpr_dt,tpr_dt,'b', label = 'DT')
+ax1.plot(fpr_lr,tpr_lr,'r', label = 'LogR')
+ax1.plot(fpr_rf,tpr_rf,'g', label = 'RF')
+ax1.set_xlabel('fpr')
+ax1.set_ylabel('tpr')
+ax1.set_title('ROC AUC Comparison', {'fontsize':15})
+ax1.legend()
+
+ax2.plot([1,0],'k--')
+ax2.plot(recall_k_8,prec_k_8,'c',label = 'KNC')
+ax2.plot(recall_dt,prec_dt,'b', label = 'DT')
+ax2.plot(recall_lr,prec_lr,'r', label = 'LogR')
+ax2.plot(recall_rf,prec_rf,'g', label = 'RF')
+ax2.set_xlabel('recall')
+ax2.set_ylabel('precision')
+ax2.set_title('ROC PRC Comparison', {'fontsize':15})
+ax2.legend();
```

Comment:
@Ettore1991 most important comment. We miss the correct baseline. If dataset is imbalanced, then baseline is not 50% anymore